File: frontend/src/api/performance.ts
/**
 * Classification performance endpoint (confusion matrix + report).
 */

import { api } from "../lib/api-client";
import type { TaxonomicRank } from "../lib/taxonomic-rank";

export interface ClassMetrics {
  class_name: string;
  common_name: string;
  scientific_name: string;
  support: number;
  precision: number | null;
  recall: number | null;
  f1: number | null;
}

export interface PerformanceResponse {
  taxonomic_rank: TaxonomicRank;
  classes: string[];
  class_scientific_names: string[];
  class_common_names: string[];
  class_taxonomy_ids: (string | null)[];
  matrix: number[][];
  row_totals: number[];
  col_totals: number[];
  grand_total: number;
  per_class: ClassMetrics[];
  macro_precision: number | null;
  macro_recall: number | null;
  macro_f1: number | null;
  weighted_precision: number | null;
  weighted_recall: number | null;
  weighted_f1: number | null;
  skipped_no_prediction: number;
  skipped_unverified: number;
  has_classifier: boolean;
  top_n_applied: number | null;
  other_bucket_present: boolean;
}

export interface PerformanceFilters {
  siteIds?: string[];
  dateFrom?: string;
  dateTo?: string;
  taxonomicRank?: TaxonomicRank;
  /** Integer or the literal 'all'. Defaults to 20 server-side. */
  topN?: string;
}

export const performanceApi = {
  get: (projectId: string, filters: PerformanceFilters = {}) => {
    const params = new URLSearchParams();
    params.set("project_id", projectId);
    if (filters.siteIds?.length) params.set("site_ids", filters.siteIds.join(","));
    if (filters.dateFrom) params.set("date_from", filters.dateFrom);
    if (filters.dateTo) params.set("date_to", filters.dateTo);
    if (filters.taxonomicRank) params.set("taxonomic_rank", filters.taxonomicRank);
    if (filters.topN) params.set("top_n", filters.topN);
    return api.get<PerformanceResponse>(
      `/api/statistics/performance?${params.toString()}`
    );
  },
};

File: backend/app/api/schemas/performance.py
"""
Schemas for the classification performance endpoint
(confusion matrix + classification report).
"""

from pydantic import BaseModel, Field


class ClassMetrics(BaseModel):
    class_name: str
    common_name: str
    scientific_name: str
    support: int
    precision: float | None
    recall: float | None
    f1: float | None


class PerformanceResponse(BaseModel):
    taxonomic_rank: str = Field(
        ...,
        description="Aggregation rank used (all / class / order / family / genus / species)",
    )
    classes: list[str] = Field(..., description="Class identifiers in matrix order")
    class_scientific_names: list[str] = Field(
        ...,
        description="Scientific labels, aligned with `classes` by index",
    )
    class_common_names: list[str] = Field(
        ...,
        description="Common labels, aligned with `classes` by index",
    )
    class_taxonomy_ids: list[str | None] = Field(
        ...,
        description=(
            "LabelTaxonomy UUID per class, aligned with `classes`. None for "
            "detector categories (animal / person / vehicle), the 'other' "
            "bucket, and any class with no direct taxonomy row."
        ),
    )
    matrix: list[list[int]] = Field(
        ...,
        description="matrix[i][j] = count of (true_class=i, predicted_class=j)",
    )
    row_totals: list[int]
    col_totals: list[int]
    grand_total: int
    per_class: list[ClassMetrics]
    macro_precision: float | None
    macro_recall: float | None
    macro_f1: float | None
    weighted_precision: float | None
    weighted_recall: float | None
    weighted_f1: float | None
    skipped_no_prediction: int = Field(
        ...,
        description=(
            "Verified detections excluded because the original prediction "
            "was unavailable (typically analysed before the prediction-"
            "history columns existed)."
        ),
    )
    skipped_unverified: int = Field(
        ...,
        description="Detections matching the filter but not yet verified.",
    )
    has_classifier: bool = Field(
        ...,
        description="False for detector-only projects.",
    )
    top_n_applied: int | None
    other_bucket_present: bool
